File: packages/cpp/examples/flow_openai_api.cpp
// docs:start provider-flow
#include "axllm/axllm.hpp"

#include <cstdlib>
#include <iostream>

int main() {
  const char* key = std::getenv("OPENAI_API_KEY");
  if (key == nullptr || std::string(key).empty()) key = std::getenv("OPENAI_APIKEY");
  if (key == nullptr || std::string(key).empty()) {
    std::cerr << "Set OPENAI_API_KEY or OPENAI_APIKEY to run this provider API example.\n";
    return 2;
  }

  const char* model = std::getenv("AX_OPENAI_MODEL");
  axllm::OpenAICompatibleClient client(axllm::object({
      {"api_key", key},
      {"model", model == nullptr || std::string(model).empty() ? "gpt-4.1-mini" : model},
      {"model_config", axllm::object({{"temperature", 0}})},
  }));
  axllm::AxGen outline = axllm::ax("topic:string -> outline:string");
  axllm::AxFlow program = axllm::flow(axllm::object({{"id", "examples.openaiApiFlow"}}))
      .execute("outline", outline)
      .map("summary",
           [](axllm::Value) {
             return axllm::object({{"summary", "Generated outline with typed Ax program steps."}});
           })
      .returns(axllm::object({{"outline", "outline"}, {"summary", "summary"}}));
  axllm::Value output = program.forward(
      client,
      axllm::object({{"topic", "how Ax composes typed LLM programs"}}));
  std::cout << axllm::stringify(output) << "\n";
}
// docs:end provider-flow

File: packages/python/examples/flow_openai_api.py
import json
import os

from axllm import OpenAICompatibleClient, ax, flow


api_key = os.getenv("OPENAI_API_KEY") or os.getenv("OPENAI_APIKEY")
if not api_key:
    raise SystemExit("Set OPENAI_API_KEY or OPENAI_APIKEY to run this provider API example.")

client = OpenAICompatibleClient(
    api_key=api_key,
    model=os.getenv("AX_OPENAI_MODEL", "gpt-5.4-mini"),
    model_config={"temperature": 0},
)
outline = ax("topic:string -> outline:string")
program = (
    flow({"id": "examples.openaiApiFlow"})
    .execute("outline", outline)
    .map("summary", lambda state: {"summary": "Generated outline with typed Ax program steps."})
    .returns({"outline": "outline", "summary": "summary"})
)
output = program.forward(client, {"topic": "how Ax composes typed LLM programs"})

print(json.dumps(output, indent=2, sort_keys=True))
